File: pt-web/src/types/tour.ts
export enum Difficulty {
  BEGINNER = "BEGINNER",
  EXPERIENCED = "EXPERIENCED",
  PRO = "PRO",
  EASY = "EASY",
}

export enum DifficultyLevel {
  EASY = "EASY",
  MEDIUM = "MEDIUM",
  HARD = "HARD",
}

export interface TourDay {
  day: number;
  plan: string;
  description: string;
  imgUrl?: string;
}

export interface FaqItem {
  id?: string;
  question: string;
  answer: string;
}

export interface TourActivity {
  activity: string;
  iconName: string;
}

export interface TourDateView {
  dateFrom: string;
  dateTo: string;
  price: number;
  description: string;
}

export interface TourView {
  id: number;
  starAmount: number;
  reviewAmount: number;
  slug: string;
  title: string;
  description: string;
  singleRoomSupplement?: number;

  location: string;
  startLocation: string;
  endLocation: string;
  durationDays: number;
  languages?: string[];
  difficulty?: Difficulty;
  minAge?: number | null;
  availableMonths?: string[];

  coverUrl: string;
  photos: string[];

  dates: TourDateView[];
  dailyItinerary?: TourDay[];
  faq: FaqItem[];

  activities: TourActivity[];
  included: string[];
  summary: string[];

  groupSize: number;
  spotsLeft: number;
  subtitle: string;

  // VIP and Room fields
  isShowVip: boolean;
  isShowRooms: boolean;
  vipPrice: number;
  roomPrice: number;

  // Popup-related fields
  popUp1Title: string;
  popUp1Description: string;
  popUp1ImageUrl: string;
  popUp2Title: string;
  popUp2Description: string;
  popUp2ImageUrl: string;

  // CTA fields for feedback block
  ctaTitle: string;
  ctaDescription: string;

  // Reviews section dynamic title
  reviewsSectionName?: string;
}

export interface TourDataFromApi {
  title: string;
  description: string;
  region: string;
  difficulty: DifficultyLevel;
  price: number | "";
  program: string | { text: string };
  dates: string | Array<{ date?: string }>;
  photos: File[];
}

export interface TourData {
  title: string;
  description: string;
  region: string;
  difficulty: DifficultyLevel;
  price: number | "";
  program: string;
  dates: string;
  photos: File[];
  videos: File[];
  coverUrl: string;
  durationDays: string;
  startLocation: string;
  endLocation: string;
  minAge: number;
  languages: string[];
  availableMonths: string[];
  reviewsSectionName: string;
  isShowVip: boolean;
  isShowRooms: boolean;
  vipPrice: number;
  roomPrice: number;
}

export interface AdminTourDate {
  id: string;
  dateFrom: string;
  dateTo: string;
  groupSize: number;
  isAvailable: boolean;
  price: number;
  description: string;
}

export interface AdminTourPhoto {
  id: string;
  url: string;
  alt?: string;
  description?: string;
}

export interface AdminTour {
  id: string;
  slug: string;
  title: string;
  description: string;
  difficulty: Difficulty;
  coverUrl: string;
  durationDays: string;
  startLocation: string;
  endLocation: string;
  location: string;
  minAge: number;
  languages: string[];
  availableMonths: string[];
  program: { days: TourDay[] };
  faq: { questions: FaqItem[] };
  activities: TourActivity[];
  included: string[];
  summary: string[];
  groupSize: number;
  spotsLeft: number;
  subtitle: string;
  popUp1Title: string;
  popUp1Description: string;
  popUp1ImageUrl: string;
  popUp2Title: string;
  popUp2Description: string;
  popUp2ImageUrl: string;
  ctaTitle: string;
  ctaDescription: string;
  reviewsSectionName: string;
  isShowVip: boolean;
  isShowRooms: boolean;
  vipPrice: number;
  roomPrice: number;
  dates: AdminTourDate[];
  photos: AdminTourPhoto[];
}

export interface UpdateTourAdminData {
  title?: string;
  slug?: string;
  description?: string;
  difficulty?: string;
  coverUrl?: string;
  durationDays?: string;
  startLocation?: string;
  endLocation?: string;
  location?: string;
  minAge?: number;
  program?: { days: TourDay[] };
  faq?: { questions: FaqItem[] };
  languages?: string[];
  availableMonths?: string[];
  activities?: TourActivity[];
  included?: string[];
  summary?: string[];
  groupSize?: number;
  spotsLeft?: number;
  subtitle?: string;
  popUp1Title?: string;
  popUp1Description?: string;
  popUp1ImageUrl?: string;
  popUp2Title?: string;
  popUp2Description?: string;
  popUp2ImageUrl?: string;
  ctaTitle?: string;
  ctaDescription?: string;
  reviewsSectionName?: string;
  isShowVip?: boolean;
  isShowRooms?: boolean;
  vipPrice?: number;
  roomPrice?: number;
  photos?: AdminTourPhoto[];
}

File: pt-general-go/internal/repository/mapper/tour.go
package mapper

import (
	db "pt-general-go/internal/db/sqlc"
	"pt-general-go/internal/domain"
)

func MapToDomainTour(row *db.Tour) *domain.Tour {
	tour := &domain.Tour{
		ID:                PgUUIDToUUID(row.ID),
		Slug:              row.Slug,
		Title:             row.Title,
		Description:       row.Description,
		Difficulty:        domain.DifficultyLevel(row.Difficulty),
		Program:           row.Program,
		FAQ:               row.Faq,
		Languages:         row.Languages,
		AvailableMonths:   row.AvailableMonths,
		CreatedAt:         row.CreatedAt.Time,
		UpdatedAt:         row.UpdatedAt.Time,
		PopUp1Title:       row.PopUp1Title,
		PopUp1Description: row.PopUp1Description,
		PopUp2Title:       row.PopUp2Title,
		PopUp2Description: row.PopUp2Description,
		PopUp1ImageUrl:    row.PopUp1ImageUrl,
		PopUp2ImageUrl:    row.PopUp2ImageUrl,
	}

	if row.Price.Valid {
		tour.Price = &row.Price.Float64
	}

	if row.StartLocation.Valid {
		tour.StartLocation = &row.StartLocation.String
	}

	if row.EndLocation.Valid {
		tour.EndLocation = &row.EndLocation.String
	}

	if row.Location.Valid {
		tour.Location = &row.Location.String
	}

	if row.DurationDays.Valid {
		tour.DurationDays = &row.DurationDays.Int32
	}

	if row.MinAge.Valid {
		tour.MinAge = &row.MinAge.Int32
	}

	if row.CoverUrl.Valid {
		tour.CoverURL = &row.CoverUrl.String
	}

	tour.GuideID = PgUUIDToUUIDPtr(row.GuideID)

	if row.GroupSize.Valid {
		tour.GroupSize = &row.GroupSize.Int32
	}

	if row.SpotsLeft.Valid {
		tour.SpotsLeft = &row.SpotsLeft.Int32
	}

	if row.Subtitle.Valid {
		tour.Subtitle = &row.Subtitle.String
	}

	return tour
}

func MapToDomainTours(rows []db.Tour) []domain.Tour {
	tours := make([]domain.Tour, 0, len(rows))
	for _, row := range rows {
		tour := domain.Tour{
			ID:                PgUUIDToUUID(row.ID),
			Slug:              row.Slug,
			Title:             row.Title,
			Description:       row.Description,
			Difficulty:        domain.DifficultyLevel(row.Difficulty),
			Program:           row.Program,
			FAQ:               row.Faq,
			Languages:         row.Languages,
			AvailableMonths:   row.AvailableMonths,
			CreatedAt:         row.CreatedAt.Time,
			UpdatedAt:         row.UpdatedAt.Time,
			PopUp1Title:       row.PopUp1Title,
			PopUp1Description: row.PopUp1Description,
			PopUp2Title:       row.PopUp2Title,
			PopUp2Description: row.PopUp2Description,
			PopUp1ImageUrl:    row.PopUp1ImageUrl,
			PopUp2ImageUrl:    row.PopUp2ImageUrl,
		}

		if row.Price.Valid {
			tour.Price = &row.Price.Float64
		}

		if row.StartLocation.Valid {
			tour.StartLocation = &row.StartLocation.String
		}

		if row.EndLocation.Valid {
			tour.EndLocation = &row.EndLocation.String
		}

		if row.Location.Valid {
			tour.Location = &row.Location.String
		}

		if row.DurationDays.Valid {
			tour.DurationDays = &row.DurationDays.Int32
		}

		if row.MinAge.Valid {
			tour.MinAge = &row.MinAge.Int32
		}

		if row.CoverUrl.Valid {
			tour.CoverURL = &row.CoverUrl.String
		}

		tour.GuideID = PgUUIDToUUIDPtr(row.GuideID)

		if row.GroupSize.Valid {
			tour.GroupSize = &row.GroupSize.Int32
		}

		if row.SpotsLeft.Valid {
			tour.SpotsLeft = &row.SpotsLeft.Int32
		}

		if row.Subtitle.Valid {
			tour.Subtitle = &row.Subtitle.String
		}

		tours = append(tours, tour)
	}
	return tours
}
